File: backend/wassily/chain/holders.py
"""Holder count replayed from ERC-20 Transfer events."""

from __future__ import annotations

from typing import Any, Protocol

from wassily.chain.abi import NATIVE, TOPICS, topic_address


class LogReader(Protocol):
    async def get_logs(self, from_block: int, to_block: int, address: Any = None, topics: Any = None) -> list[dict]:
        ...


def replay_balances(logs: list[dict[str, Any]]) -> dict[str, int]:
    balances: dict[str, int] = {}
    for entry in logs:
        topics = entry.get("topics") or []
        data = entry.get("data", "0x")
        if len(topics) != 3 or len(data) < 66:
            continue  # ERC-721 transfers index the id instead
        value = int(data[2:66], 16)
        sender = topic_address(topics[1])
        receiver = topic_address(topics[2])
        if sender != NATIVE:
            balances[sender] = balances.get(sender, 0) - value
        if receiver != NATIVE:
            balances[receiver] = balances.get(receiver, 0) + value
    return balances


async def holders_at(rpc: LogReader, token: str, from_block: int, to_block: int) -> int:
    """Addresses with a positive balance at ``to_block``. Pools and contracts count, as on a block explorer."""
    logs = await rpc.get_logs(from_block, to_block, token, [TOPICS.transfer])
    return sum(1 for balance in replay_balances(logs).values() if balance > 0)

File: src/server/chain/holders.ts
import { NATIVE, TOPICS, topicAddress } from './abi';
import type { RpcClient } from './rpc';

/**
 * Holder count at `toBlock`: replay every ERC-20 Transfer of `token` since
 * `fromBlock` and count addresses with a positive balance. Pools and
 * contracts count as holders, the same as on a block explorer.
 * Returns null when the replay is incomplete: an address that sent more than it
 * received means transfers before `fromBlock` were missed, so no count is honest.
 */
export async function holdersAt(rpc: Pick<RpcClient, 'getLogs'>, token: string, fromBlock: number, toBlock: number): Promise<number | null> {
  const logs = await rpc.getLogs({ fromBlock, toBlock, address: token, topics: [TOPICS.transfer] });
  const balances = new Map<string, bigint>();
  for (const log of logs) {
    if (log.topics.length !== 3 || log.data.length < 66) continue; // ERC-721 transfers index the id instead
    const value = BigInt(log.data.slice(0, 66));
    const from = topicAddress(log.topics[1]);
    const to = topicAddress(log.topics[2]);
    if (from !== NATIVE) balances.set(from, (balances.get(from) ?? 0n) - value);
    if (to !== NATIVE) balances.set(to, (balances.get(to) ?? 0n) + value);
  }
  let holders = 0;
  for (const balance of balances.values()) {
    if (balance < 0n) return null;
    if (balance > 0n) holders++;
  }
  return holders;
}
